### PR TITLE
feat(codecommit): make Repository a source for CodeStar Notifications

### LINKED ISSUE #15653: (@aws-cdk/aws-codestarnotifications): CodeCommit repository not supported as source for Notification Rule
Currently @aws-cdk/aws-codestarnotifications module does not seem to support using repository as the source. However, this does appear to be supported in Cloudformation.
From the CDK api docs:
["Currently, Supported sources include pipelines in AWS CodePipeline and build projects in AWS CodeBuild in this L2 constructor."](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-codestarnotifications.NotificationRuleProps.html#source)

and from CloudFormation template reference:
["Supported resources include pipelines in AWS CodePipeline, repositories in AWS CodeCommit, and build projects in AWS CodeBuild."](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codestarnotifications-notificationrule.html#aws-resource-codestarnotifications-notificationrule-properties)

In the console notification rules can easily be created for repository events and sent to SNS or Chatbot targets. I would like to replicate that with the CDK. 

### Environment

  - **CDK CLI Version:** 1.109.0
  - **Module Version:** 1.109.0
  - **Node.js Version:** 14.17.0
  - **OS:** Windows 10
  - **Language (Version):** 4.3.4

### Other information
According to CloudFormation, an ARN is required, however re-creating an ARN won't work `INotificationRuleSource` won't accept a string. Ideally it should work something like this:

```
    const repositoryNotification = new NotificationRule(this, 'PullRequestNotificationRule', {
      source: repository.repositoryArn,
      events: [
        'codecommit-repository-pull-request-created',
        'codecommit-repository-pull-request-merged'
      ],
      targets: [chatbot],
      detailType: DetailType.FULL,
      notificationRuleName: 'pull-request-status'
    })
    repositoryNotification.addTarget(chatbot)
```

### PR DESCRIPTION
**Pull request by badfun**
----
Fixes #15653

Notification rule can be created for CodeCommit Repository events using @aws-cdk/aws-codestarnotifications

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by badfun** [comment]
@luckily @skinny85 - I have re-submitted the pull request for [issue #15653](https://github.com/aws/aws-cdk/issues/15653) here.

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 1]
```diff
@@ -110,8 +123,128 @@ export interface IRepository extends IResource {
    * Grant the given identity permissions to read this repository.
    */
   grantRead(grantee: iam.IGrantable): iam.Grant;
+
+
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:127]: One empty line too many 🙂.

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 2]
```diff
@@ -110,8 +123,128 @@ export interface IRepository extends IResource {
...
+  /**
+   *
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:129]: No need for this empty line.

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 3]
```diff
@@ -110,8 +123,128 @@ export interface IRepository extends IResource {
...
+  notifyOn(
+    id: string,
+    target: notifications.INotificationRuleTarget,
+    options: RepositoryNotifyOnOptions,
+  ): notifications.INotificationRule;
+
+  /**
+   * Defines a CodeStar Notification rule which triggers when a comment is made on a commit.
+   */
+  notifyOnCommentOnCommits(
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:150]:
```suggestion
  notifyOnCommitComment(
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 4]
```diff
@@ -110,8 +123,128 @@ export interface IRepository extends IResource {
...
+    id: string,
+    target: notifications.INotificationRuleTarget,
+    options?: notifications.NotificationRuleOptions,
+  ): notifications.INotificationRule;
+
+  /**
+   * Defines a CodeStar Notification rule which triggers when a comment is made on a pull request
+   */
+  notifyOnCommentOnPullRequests(
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:159]:
```suggestion
  notifyOnPullRequestsComment(
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 5]
```diff
@@ -110,8 +123,128 @@ export interface IRepository extends IResource {
...
+    id: string,
+    target: notifications.INotificationRuleTarget,
+    options?: notifications.NotificationRuleOptions,
+  ): notifications.INotificationRule;
+
+  /**
+   * Defines a CodeStar Notification rule which triggers when an approval status is changed
+   */
+  notifyOnApprovalsStatusChanged(
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:168]:
```suggestion
  notifyOnApprovalStatusChanged(
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 6]
```diff
@@ -110,8 +123,128 @@ export interface IRepository extends IResource {
...
+    id: string,
+    target: notifications.INotificationRuleTarget,
+    options?: notifications.NotificationRuleOptions,
+  ): notifications.INotificationRule;
+
+  /**
+   * Defines a CodeStar Notification rule which triggers when an approval rule is overridden
+   */
+  notifyOnApprovalsRuleOverride(
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:177]:
```suggestion
  notifyOnApprovalRuleOverridden(
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 7]
```diff
@@ -110,8 +123,128 @@ export interface IRepository extends IResource {
...
+  notifiyOnPullRequestMerged(
+    id: string,
+    target: notifications.INotificationRuleTarget,
+    options?: notifications.NotificationRuleOptions,
+  ): notifications.INotificationRule;
+
+  /**
+   * Defines a CodeStar Notification rule which triggers when a new branch or tag are created
+   */
+  notifyOnBranchesAndTagsCreated(
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:222]:
```suggestion
  notifyOnBranchOrTagCreated(
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 8]
```diff
@@ -110,8 +123,128 @@ export interface IRepository extends IResource {
...
+    id: string,
+    target: notifications.INotificationRuleTarget,
+    options?: notifications.NotificationRuleOptions,
+  ): notifications.INotificationRule;
+
+  /**
+   * Defines a CodeStar Notification rule which triggers when a branch or tag is deleted
+   */
+  notifyOnBranchesAndTagsDeleted(
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:231]:
```suggestion
  notifyOnBranchOrTagDeleted(
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 9]
```diff
@@ -110,8 +123,128 @@ export interface IRepository extends IResource {
...
+    id: string,
+    target: notifications.INotificationRuleTarget,
+    options?: notifications.NotificationRuleOptions,
+  ): notifications.INotificationRule;
+
+  /**
+   * Defines a CodeStar Notification rule which triggers when a branch or tag is updated
+   */
+  notifyOnBranchesAndTagsUpdated(
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:240]:
```suggestion
  notifyOnBranchOrTagUpdated(
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 10]
```diff
@@ -110,8 +123,128 @@ export interface IRepository extends IResource {
...
+    id: string,
+    target: notifications.INotificationRuleTarget,
+    options?: notifications.NotificationRuleOptions,
+  ): notifications.INotificationRule;
 }
 
+
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:247]: No need for this new empty line.

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 11]
```diff
@@ -110,8 +123,128 @@ export interface IRepository extends IResource {
...
+  /**
+   * Defines a CodeStar Notification rule which triggers when a branch or tag is updated
+   */
+  notifyOnBranchesAndTagsUpdated(
+    id: string,
+    target: notifications.INotificationRuleTarget,
+    options?: notifications.NotificationRuleOptions,
+  ): notifications.INotificationRule;
 }
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:245]: This is way too many of these new methods. Are really all of them useful? How about we start with 2-3, plus the generic `notifyOn()` one? We can always add them later as needed.

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 12]
```diff
@@ -110,8 +123,128 @@ export interface IRepository extends IResource {
...
+   *
+   * Defines a CodeStar Notification rule triggered when the project
+   * events specified by you are emitted. Similar to `onEvent` API.
+   *
+   * You can also use the methods to define rules for the specific event emitted.
+   * eg: `notifyOnPullRequstCreated`
+   *
+   * @param id
+   * @param target
+   * @param options
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:138]: Either fill these, or remove them (they're not useful empty).

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 13]
```diff
@@ -268,6 +401,152 @@ abstract class RepositoryBase extends Resource implements IRepository {
...
     );
   }
+
+  /**
+   * Defines a CodeStar Notification rule
+   *
+   * @param id
+   * @param target
+   * @param options
+   */
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:411]: Don't repeat the JSDocs here, there's no need.

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 14]
```diff
@@ -448,3 +727,64 @@ function makeCloneUrl(stack: Stack, repositoryName: string, protocol: 'https' |
...
+
+/**
+ * List of event types for AWS CodeCommit
+ * @see https://docs.aws.amazon.com/dtconsole/latest/userguide/concepts.html#events-ref-repositories
+ */
+export enum RepositoryNotificationEvents {
+  /**
+   * Trigger notication when comment made on commit
+   */
+  COMMENT_ON_COMMITS = 'codecommit-repository-comments-on-commits',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:739]:
```suggestion
  COMMIT_COMMENT = 'codecommit-repository-comments-on-commits',
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 15]
```diff
@@ -448,3 +727,64 @@ function makeCloneUrl(stack: Stack, repositoryName: string, protocol: 'https' |
...
+
+  /**
+   * Trigger notification when comment made on pull request
+   */
+  COMMENT_ON_PULL_REQUESTS = 'codecommit-repository-comments-on-pull-requests',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:744]:
```suggestion
  PULL_REQUEST_COMMENT = 'codecommit-repository-comments-on-pull-requests',
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 16]
```diff
@@ -448,3 +727,64 @@ function makeCloneUrl(stack: Stack, repositoryName: string, protocol: 'https' |
...
+
+  /**
+   * Trigger notification when approval status changed
+   */
+  APPROVALS_STATUS_CHANGED = 'codecommit-repository-approvals-status-changed',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:749]:
```suggestion
  APPROVAL_STATUS_CHANGED = 'codecommit-repository-approvals-status-changed',
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 17]
```diff
@@ -448,3 +727,64 @@ function makeCloneUrl(stack: Stack, repositoryName: string, protocol: 'https' |
...
+
+  /**
+   * Trigger notifications when approval rule is overridden
+   */
+  APPROVALS_RULE_OVERRIDE = 'codecommit-repository-approvals-rule-override',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:754]:
```suggestion
  APPROVAL_RULE_OVERRIDDEN = 'codecommit-repository-approvals-rule-override',
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 18]
```diff
@@ -448,3 +727,64 @@ function makeCloneUrl(stack: Stack, repositoryName: string, protocol: 'https' |
...
+
+  /**
+   * Trigger notification when pull requset is merged
+   */
+  PULL_REQUEST_MERGED = 'codecommit-repository-pull-request-merged',
+
+  /**
+   * Trigger notification when a branch or tag is created
+   */
+  BRANCHES_AND_TAGS_CREATED = 'codecommit-repository-branches-and-tags-created',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:779]:
```suggestion
  BRANCH_OR_TAG_CREATED = 'codecommit-repository-branches-and-tags-created',
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 19]
```diff
@@ -448,3 +727,64 @@ function makeCloneUrl(stack: Stack, repositoryName: string, protocol: 'https' |
...
+
+  /**
+   * Trigger notification when a branch or tag is deleted
+   */
+  BRANCHES_AND_TAGS_DELETED = 'codecommit-repository-branches-and-tags-deleted',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:784]:
```suggestion
  BRANCH_OR_TAG_DELETED = 'codecommit-repository-branches-and-tags-deleted',
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 20]
```diff
@@ -1,10 +1,23 @@
+import * as notifications from '@aws-cdk/aws-codestarnotifications';
 import * as events from '@aws-cdk/aws-events';
 import * as iam from '@aws-cdk/aws-iam';
 import { IResource, Lazy, Resource, Stack } from '@aws-cdk/core';
 import { Construct } from 'constructs';
 import { CfnRepository } from './codecommit.generated';
 
-export interface IRepository extends IResource {
+/**
+ * Additional options to pass to the notification rule
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:9]: Docs should be complete sentences, with fullstops.

Same comment everywhere in the PR.

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts [hunk 21]
```diff
@@ -448,3 +727,64 @@ function makeCloneUrl(stack: Stack, repositoryName: string, protocol: 'https' |
...
+
+  /**
+   * Trigger notification when a branch or tag is deleted
+   */
+  BRANCHES_AND_TAGS_DELETED = 'codecommit-repository-branches-and-tags-deleted',
+
+  /**
+   * Trigger notification when a branch or tag is updated
+   */
+  BRANCHES_AND_TAGS_UPDATED = 'codecommit-repository-branches-and-tags-updated',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/lib/repository.ts:789]:
```suggestion
  BRANCH_OR_TAG_UPDATED = 'codecommit-repository-branches-and-tags-updated',
```

File: packages/@aws-cdk/aws-codecommit/package.json [hunk 22]
```diff
@@ -98,7 +99,8 @@
     "@aws-cdk/aws-events": "0.0.0",
     "@aws-cdk/aws-iam": "0.0.0",
     "@aws-cdk/core": "0.0.0",
-    "constructs": "^3.3.69"
+    "constructs": "^3.3.69",
+    "@aws-cdk/aws-codestarnotifications": "0.0.0"
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/package.json:103]: Please put this at the beginning of the list, like you did for `"dependencies"`.

**Review by skinny85 — DISMISSED** [review]
Looks great @badfun, thanks for the contribution!

A few comments, mainly around naming.

**Comment by badfun** [comment]
Thanks @skinny85,
I will make the changes asap.

**Comment by skinny85** [comment]
@badfun  please re-request my review (there's a button in the top-right of the PR window, next to my avatar) once you're ready for another round of reviews. Thanks for the contribution!

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts (hunk 5 above)
Comment (badfun) [reply packages/@aws-cdk/aws-codecommit/lib/repository.ts:168]: Shouldn't this stay plural, to match the event itself: 'codecommit-repository-approvals-status-changed'?

I prefer the singular myself for camel-case, but the event is plural.

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts (hunk 9 above)
Comment (badfun) [reply packages/@aws-cdk/aws-codecommit/lib/repository.ts:240]: Again, shouldn't this match the event? It is 'codecommit-repository-branches-and-tags-updated'. 
"and" not "or".

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts (hunk 11 above)
Comment (badfun) [reply packages/@aws-cdk/aws-codecommit/lib/repository.ts:245]: I can't answer what is useful to other users. I was just matching what is available in the console. For me, I just needed the pull request methods. I can certainly cut some out, and then `notifyOn()` could be used for the others.

**Comment by badfun** [comment]
@skinny85 I've made most of the changes. Looking for some clarification on a few points commented above.
thanks,
Ken

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts (hunk 4 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codecommit/lib/repository.ts]: Apologies, this should actually be

```suggestion
  notifyOnPullRequestComment(
```

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts (hunk 5 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codecommit/lib/repository.ts:163]: I'm fine using a different name.

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts (hunk 9 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codecommit/lib/repository.ts:235]: I'm fine using a different name.

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts (hunk 20 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codecommit/lib/repository.ts:9]: Still not resolved I see 🙂.

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts (hunk 11 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codecommit/lib/repository.ts:240]: Yes, let's cut some of the more "esoteric" ones (`notifyOnPullRequestSourceUpdated()`?).

File: packages/@aws-cdk/aws-codecommit/lib/repository.ts (hunk 10 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codecommit/lib/repository.ts:242]: Still here.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Let's change the names according to my comments (I'm fine with the fact that they are not identical to the event names).

**Comment by badfun** [comment]
@skinny85  I believe I got everything now. Let me know if there is something I've missed.

**Pull request merged**